onyo/lib/onyo.py

```diff
@@ -1240,3 +1241,128 @@ def unset(self, paths: Iterable[Union[Path, str]], values: str,
             self.restore()
 
         return diff
+
+    #
+    # GET
+    #
+    # TODO: This should be a staticmethod in utils
+    def _walk(self, path, depth) -> Generator:
+        """Recursively yield filepaths up to a given depth"""
+        depth -= 1
+        with os.scandir(path) as p:
+            for entry in p:
+                # ignore hidden files and folders
+                # TODO: Better to use _is_protected_path or ignore hidden?
+                #   `if self._is_protected_path(entry.path):`
+                if entry.name.startswith('.'):
+                    continue
+
+                # yield discovered assets
+                if not entry.is_dir():
+                    yield entry.path
+
+                # explore dirs if depth allows
+                if entry.is_dir() and depth >= 0:
+                    yield from self._walk(entry.path, depth)
+
+    @staticmethod
+    def _filter_by_name(assets: List[str], filters: dict) -> Generator:
+        """Filter assets by filename to prevent redundant asset loading"""
+        keys = ['type', 'make', 'model', 'serial']
+        for asset in assets:
+            split = set(zip(keys, re.split('[_.]', os.path.basename(asset))))
+            if set(filters.items()).issubset(split):
+                yield asset
+
+    @staticmethod
+    def _get_asset_data(
+            assets: Union[Generator[Any, Any, Any], List[Any]], keys: Set[str],
+            filters: Union[dict, None] = None) -> Generator:
+        """Return data from asset keys for assets matching filters"""
+        yaml = YAML(typ='safe')
+        default_keys = ['type', 'make', 'model', 'serial']
+        string_types = {  # TODO: Do we have a list of accepted string types?
```

Comment:
It sounds like you're proposing a syntax to filter field values with typing.

The use case for this is unclear to me. Invalid fields should be handled by the (future) validation system.

> My string_types listing is flawed, however, because '2' can be a valid string. Instead, we should use [num] and [alpha] to be explicit in what the type means. 

I strongly disagree with this. `'2'` is a string. Attempting to cast the contents into another datatype and declaring it as that other datatype is a slow slide into chaos. At that point, `country: "NO"` is a boolean, and will never be treated as a string.

> for now I will limit the list to [unset], [list] and [dict] only. This can be expanded upon, but not within the scope of this PR.

I agree that another PR should more fully flesh out the filtering system. We touched on some of those complexities in previous conversations. I'm comfortable with `<unset>`, `<list>`, and `<dict>` for now. :-)